### ISSUE TITLE
near-vm: the `LimitedMemoryPool` should allocate an appropriate number of memories for the maximum concurrency

### ISSUE THREAD

**Issue by nagisa** [near/nearcore#10732]
**Describe the bug**

Currently the LimitedMemoryPool that's constructed for near-vm hardcodes the number of memories that the pool contains. However this is not exactly the right way to go about it, especially as the transaction runtime can be configured to result in an arbitrary number of concurrent operations -- e.g. view_client_thread or num_shards options.

The transaction runtime should set up the runtime with the maximum concurrency it expects, and the contract runtime should enforce this concurrency limit by blocking (semaphores, condvars) rather than failing.

**To Reproduce**

Invoke the contract runtime with a huge concurrency.

**Expected behavior**

No failures should occur.

**Comment by akhi3030** [comment]
Do I understand correctly that a single `LimitedMemoryPool` is used to serve all memory requests from all contract executions?  Why do we need to limit this pool in the first place?

**Comment by nagisa** [comment]
> Do I understand correctly that a single LimitedMemoryPool is used to serve all memory requests from all contract executions?

A single contract execution needs exactly one memory from this pool. It is populated with the contract machine code. The number of memories *may* increase to two in the future, the second being used to serve the contract's working memory, but it most likely would be served by a second pool instead.

> Why do we need to limit this pool in the first place?

We should *not* allow an unbounded resource growth (think: memory leak) for the contract runtime, while also providing performance characteristics that are predictable. The requirement for predictable performance means pre-allocating or otherwise reusing the resources we need ahead of time to the best of our estimate.

If we didn't care about the unbounded growth, the implementation of would be quite straightforward, instead of freeing the resources back to the OS, we'd put them into a reuse buffer. Whenever the resource is necessary again, we would fetch it from that buffer, or if its empty we would ask OS for some new resources, while paying the one-time sacrifice in predictability. In this sort of implementation anybody who sees a very transient load peak of say $VERYMANY view calls, they would end up with an unreasonably large reuse buffer that's effectively equivalent to a leak.

The next most straightforward implementation that allows for pre-allocated resources while at the same time preventing an effectively unbounded resource leak is a resource pool that has an upper limit of the resources it can hold. That's what we have implemented right now.

This idea can certainly be iterated upon further. For example we could still go ask the OS for the necessary resources whenever there's a peak over what the preallocated resource pool can serve. But in the end this is adding complexity and tradeoffs to what could otherwise been good enough™ (given the requirements I've been aware at the time this was implemented.)

---

To add to this, historically we have had contract runtime being blamed multiple times for causing memory exhaustion leading to a node crash, often incorrectly so. A strictly bounded preallocated resource pool has a side benefit of ensuring that the resources needed by the runtime are predictable to a certain extent, which should make investigating the OOM faults easier. It also helps us to better establish answers to questions like "what are the hardware requirements to run a node".

---

Does that help to establish the necessary context?
